### PR TITLE
chore: Setup AI agent infrastructure

### PR DESCRIPTION
**Pull request by voidvore**
## Summary

This PR introduces comprehensive AI agent support infrastructure for the asusctl project. To streamline AI-assisted development workflows across modern agentic tools (Claude, Cursor, OpenCode, Antigravity, etc.), it establishes the standard .agents/ workspace directory, sub-agents, and skill files.

## Roadmap

- [x] Create CONTRIBUTING_AGENTS.md (Research-backed human governance guide for maintaining AGENTS.md)
- [x] Edit CONTRIBUTING.md
- [ ] Populate AGENTS.md
- [ ] Add sub-agents (.agents/agents/)
- [ ] Add skill files (.agents/skills/)
  - [x] Release Note skill
- [ ] Add `.coderabbit.yaml`

### CODE REVIEW

**Comment by luytan** [comment]
@coderabbitai full review

**Comment by voidvore** [comment]
@scardracs for the "Git & Workflow Guidelines" part, i think it'd be better to put it in `CONTRIBUTING.md` and tell `AGENTS.md` to read that file. i want to put things in files that both humans and llms can use together as much as possible, to reduce the development burden.

**Comment by voidvore** [comment]
@coderabbitai emit path instructions

File: .agents/AGENTS.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+| rog-aura | `rog-aura/` |
+| rog-profiles | `rog-profiles/` |
+| rog-control-center | `rog-control-center/` |
+| rog-slash | `rog-slash/` |
+| rog-scsi | `rog-scsi/` |
+| simulators | `simulators/` |
+| data (service files, udev rules) | `data/` |
+| extra (icons, web assets) | `extra/` |
+| distro-packaging | `distro-packaging/` |
+| CI/CD (GitHub Actions) | `.github/` |
```
[the remark this answers is not in the document]
Comment (voidvore) [reply .agents/AGENTS.md]: Outdated. The table now has three populated columns (Area | Location | Description), fixed in dbd4ff19. Note also that `AGENTS.md` was moved to the repo root, so this path (`.agents/AGENTS.md`) no longer exists.

File: .agents/CONTRIBUTING_AGENTS.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+2. [Research Findings](#research-findings)
+   - [1. Instruction File Structure & Impact (Instructions-as-Code)](#1-instruction-file-structure--impact-instructions-as-code)
+   - [2. Rule Polarity & Context Priming (Guardrails vs. Guidance)](#2-rule-polarity--context-priming-guardrails-vs-guidance)
+   - [3. Repository Context Overhead & Minimalist Design (AGENTS.md Evaluation)](#3-repository-context-overhead--minimalist-design-agentsmd-evaluation)
+   - [4. Closed-Loop Review Feedback & Behavioral Rule Accumulation (Self-Improving Agents)](#4-closed-loop-review-feedback--behavioral-rule-accumulation-self-improving-agents)
+3. [Contributor Conclusions](#contributor-conclusions)
+   - [Sub-Agent Usage](#sub-agent-usage)
+   - [Performance and Flow Configuration](#performance-and-flow-configuration)
+   - [Documentation File Method](#documentation-file-method)
+4. [Note to Maintainers: AI Feedback Loop](#note-to-maintainers-ai-feedback-loop)
```
[the remark this answers is not in the document]
Comment (voidvore) [reply .agents/CONTRIBUTING_AGENTS.md]: Outdated. I verified all 11 ToC anchors against GitHub's markdown render API (`POST /markdown`) and every one resolves correctly to its emoji-prefixed heading, including the leading-hyphen cases (`#-why-this-file-exists`, `#-research-findings`, `#-contributor-conclusions`, `#-note-to-maintainers-ai-feedback-loop`) and `#sub-agent-usage-and-setup-guidelines`. No changes needed.

**Comment by scardracs** [comment]
@voidvore new version released

# Asusctl Repository & Agent Governance Rules

This document outlines the strict guidelines, coding standards, daemon architecture rules, safety constraints, and Git workflow policies required when working on the `asusctl` repository.

> [!IMPORTANT]
> **Core Principle (Senior Rust Engineer Mindset):** Code quality, memory safety, and daemon stability ALWAYS take absolute precedence over implementation speed or temporary workarounds. Never passively accept assumptions, instructions, or automated review findings: proactively validate them against Rust memory/concurrency models and Linux kernel semantics. If a proposed change is flawed, redundant, suboptimal, or introduces architectural risks, explicitly explain **why** and propose the sound, production-grade approach before proceeding.

---

## 1. Git & Workflow Guidelines

### 1.1. Never Bypass Git Hooks
* **Strict Enforcement:** Do **NOT** use `--no-verify` or `-n` flags with `git commit` or `git push`.
* **Requirement:** All commits and pushes must allow standard git hooks (`pre-commit` and `pre-push`) to execute and pass cleanly.

### 1.2. Explicit Push Authorization
* **No Auto-Push:** Never execute `git push` without explicit, real-time user approval.
* **Single Event Scope:** Approval applies strictly to that specific push event. Subsequent pushes require new, dedicated confirmation.

### 1.3. Clean Local Checks
Before creating any commit, the codebase **must** pass all local verification checks specified in Section 5.1 with zero errors and zero warnings.

> [!IMPORTANT]
> Every single commit must maintain flawless formatting and zero warnings across all workspace crates.

### 1.4. Atomic Commits & Standardized Commit Messages
* **Independently Buildable & Logical Commits:** Each commit in the repository history must represent a single, self-contained, logical unit of work that is independently buildable and passes all local verification checks (Section 5.1).
* **No Trailing Fix Commits:** When bringing fixes, formatting adjustments, or lint resolutions for changes introduced in a feature branch, do **NOT** create trailing fix commits (e.g., `fix text`, `fix clippy`). Amend or rebase the fix directly into the relevant original commit so the git history consists strictly of clean, logical feature/refactor commits.
* **Distinct Commit Scope:** Do not mix unrelated refactors, formatting-only changes, and feature modifications in the same commit.
* **Conventional Commit Standard:** Commit messages must strictly follow Conventional Commits formatting (e.g., `feat:`, `fix:`, `refactor:`, `docs:`, `chore:`) to support automated `CHANGELOG.md` generation.

### 1.5. CI Requirements & Pull Request Enforcement
* **Strict Equivalence with Local Checks:** All pull requests and commits must pass 100% of CI checks (builds, unit tests, clippy, cranky, formatting).
* **No Bypassing CI Failures:** Never merge or approve code with failing CI status checks.

### 1.6. Maintainer Review Boundaries
* **Specialized Review Required:** Changes affecting D-Bus interfaces, kernel interactions (`asus-wmi`, sysfs, ioctls), root privilege boundaries, or configuration schemas require explicit maintainer review.

### 1.7. Mandatory Branch Verification
* **Always Check Active Branch:** Before analyzing, editing files, running tests, or staging commits, **ALWAYS** check and verify the active Git branch (e.g. via `git branch` or `git status`). Ensure all code modifications and context are strictly scoped to the active branch and current task.

---

## 2. Rust Safety & Code Quality Rules

### 2.1. Prohibition of `.unwrap()`
* **No Blind Unwraps:** No `.unwrap()` calls in production daemon and library runtime paths to prevent runtime panics.
* **Preferred Handling:**
  - Use proper error propagation with the `?` operator and `RogError` / custom domain errors.
  - Use pattern matching (`match`, `if let`) or safe fallback combinators (`unwrap_or`, `unwrap_or_else`, `ok_or`).
  - If a value is guaranteed to be present by prior invariant checks, use `.expect("Clear explanation of why this cannot fail")` instead of `.unwrap()`.

### 2.2. Strict Control of `unsafe` Code
* **Default Stance:** Avoid `unsafe` blocks whenever safe Rust abstractions or standard library utilities can achieve the same result.
* **Isolation Behind Safe Abstractions:** Unsafe code must be isolated behind minimal, safe public wrapper APIs whenever possible.
* **FFI & Kernel Interactions:** If an `unsafe` block is strictly necessary (e.g., raw C FFI calls or kernel ioctls), it must strictly follow these rules:
  1. Minimize the scope of the `unsafe` block to only the specific unsafe call.
  2. **Mandatory Safety Documentation:** Every `unsafe` block **MUST** be immediately preceded by a detailed `// SAFETY:` comment explaining why the invariants hold and why the operation is memory-safe and sound.

### 2.3. Extended Linting with `cargo cranky`
* Code changes must satisfy the strict lint set configured in `Cranky.toml`.
* Run `cargo cranky` to catch extra lints such as mutex integrity, unwanted `todo!` or `unimplemented!` macros, string allocation inefficiencies, and unhandled async locks.

### 2.4. Structured Logging Requirements & Prohibition of `println!`
* **No Direct Standard Output Printing:** Avoid `println!` or `eprintln!` for status, warnings, or error reporting across daemon, library, and CLI modules.
* **Structured Severity Levels:**
  - `error!`: Unrecoverable failures requiring immediate attention.
  - `warn!`: Recoverable degraded states (e.g. missing optional sysfs feature, non-fatal WMI failure).
  - `info!`: Key lifecycle events (daemon startup, controller initialization, mode changes).
  - `debug!` / `trace!`: Verbose diagnostic details and raw packet/event tracing.
* **Diagnostic Context:** Logs must include sufficient context (paths, device IDs, error tracebacks) to diagnose failures without exposing sensitive data.

### 2.5. Dependency Management & `Cargo.lock` Policy
* **Justification Required:** Do NOT add new third-party crate dependencies to `Cargo.toml` without explicit technical justification.
* **Prefer Workspace Dependencies:** Always check if existing workspace crates or standard library utilities can fulfill the requirement before adding a new crate.
* **Cargo.lock Integrity:** Do not edit `Cargo.lock` manually. Dependency updates must be intentional and reviewed to avoid unnecessary version churn.

### 2.6. MSRV (1.85) & Rust Edition 2024 Compatibility
* **MSRV (Rust 1.85) & Edition 2024 Adherence:** All workspace crates target **Rust Edition 2024** and a Minimum Supported Rust Version of **1.85** (`rust-version = "1.85"` in `Cargo.toml`).
* **Idiomatic Modern Code Updates:** When writing, updating, or refactoring code, follow idiomatic Rust Edition 2024 patterns and standard library features strictly supported up to Rust 1.85.
* **No Bleeding-Edge Incompatibilities:** Avoid using language features or APIs stabilized in Rust 1.86+ or nightly-only features without explicit justification and an agreed MSRV bump.

### 2.7. Compiler Upgrade Warning Policy
* **Zero Warning Policy Across Compiler Versions:** New compiler warnings introduced by Rust toolchain upgrades must be resolved promptly at the root cause, rather than suppressed via `#![allow(...)]` or ignored.

### 2.8. Panic Prevention & Assertion Controls
* **Production Panic Prohibition:** Production daemon and library runtime code must not intentionally panic during normal or degraded operation.
* **Restricted Macros:** Avoid `panic!`, `unreachable!`, `todo!`, `unimplemented!`, `assert!`, or `debug_assert!` in production runtime code unless strictly justified by an impossible internal invariant and documented with an explicit comment explaining why the condition cannot be reached.

---

## 3. Daemon & Architecture Design Patterns

### 3.1. Daemon Controller Pattern
* All controller components in `asusd` must implement standard daemon traits where applicable:
  - `GetSupported`: Checks if hardware/kernel features required by the controller are available.
  - `Reloadable`: For controllers requiring state reload on startup or signal.
  - `CtrlTask`: For background task loops, logind event handling (suspend/resume/boot), and file watching.
  - `ZbusAdd`: For exposing controller interfaces to the D-Bus system bus via `zbus`.

### 3.2. Concurrency, Locking Safety & Tokio Mailbox (Actor Pattern)
* **Concurrency Primitive Selection:**
  - **Tokio Mailbox / Actor Pattern (`mpsc` + `oneshot`):** Preferred for controllers managing complex, sequential hardware I/O (e.g. Aura RGB USB HID, AniMe Matrix frame loops, or udev/logind event queues). Ensures single-ownership of hardware state, serializes I/O execution, and eliminates deadlocks by construction.
  - **Shared Memory Async Locks (`Arc<RwLock<T>>` / `Arc<Mutex<T>>`):** Preferred for lightweight in-memory data structures with high-frequency reads and infrequent writes (e.g. `Config`, platform profile lookups, or small shared maps like `queued_gpu`) to minimize channel allocation and context-switch overhead.
* **Async Mutex & Lock Safety:** When sharing controllers across tasks or D-Bus handlers via shared state, always use async-compatible locks (`tokio::sync::Mutex` / `tokio::sync::RwLock`).
* **Deadlock Prevention in Task Loops:** 
  - Avoid blocking execution in event dispatch loops. 
  - Use non-blocking `try_lock()` inside task event callbacks (e.g., `self.create_sys_event_tasks(...)`) to prevent thread deadlocks when system events fire concurrently.

### 3.3. D-Bus Interface Standards & Backward Compatibility (`zbus`)
* Implement D-Bus methods on dedicated wrapper structs (e.g., `Ctrl<Name>Zbus`).
* Use `Self::add_to_server_helper` under standard object paths (e.g., `/org/asuslinux/<Feature>`).
* **Backward Compatibility:** Preserve existing D-Bus signatures and object paths so external clients (`rog-control-center`, GNOME extensions, `asusctl` CLI) do not break.

### 3.4. Kernel & Sysfs Attribute Resilience across ASUS Laptop Models
* **Extensive ASUS Hardware Diversity:** ASUS produces a vast array of laptop series (ROG Strix, Zephyrus, TUF, ZenBook, Vivobook, ProArt, Ally) with varying ACPI/WMI tables, EC registers, sysfs attributes, and kernel driver support across Linux releases.
* **No Single-Machine Invariant Assumptions:** Never assume that a hardware feature, sysfs path, or ACPI WMI call present on one test machine exists or behaves identically on all ASUS laptops.
* **Dynamic Probing & Graceful Feature Degradation:** Always check path availability dynamically at runtime (`GetSupported`, non-blocking file checks, udev enumeration). If a hardware capability is missing or unsupported on a given model, degrade gracefully (or fallback to virtual software equivalents where applicable) without crashing the daemon or emitting spurious fatal errors.
* **Strict Input/Output Parsing:** Parse and validate all strings read from or written to sysfs/debugfs nodes before casting to Rust enums or numeric types.

### 3.5. Safe Configuration Schema & Fallbacks
* When modifying configuration structs (in `config-traits` or `/etc/asusd/`), always provide sensible default implementations (`Default` trait).
* **Configuration Migration Safety:** Configuration schema migrations must be reversible where practical. Failed migrations must leave existing user configurations intact. Never silently delete or overwrite user settings.

### 3.6. D-Bus Input Validation & Security (`zbus`)
* **Never Trust External D-Bus Clients:** Validate all input parameters received via D-Bus methods (strings, numeric ranges, enum variants) before processing.
* **Structured D-Bus Error Responses:** Return explicit, serializable D-Bus error types (`zbus::fdo::Error`) with descriptive messages rather than panicking or swallowing errors silently.

### 3.7. Privilege Boundary & Execution Limits (`asusd`)
* **Privileged Daemon Boundary:** Because `asusd` runs with root privileges, strictly limit operations requiring elevated rights.
* **No Arbitrary User Command Execution:** Never execute untrusted or user-supplied shell commands directly.
* **Path Validation & Traversal Prevention:** Validate all filesystem paths read or written by the daemon. Restrict operations to explicit sysfs, debugfs, `/etc/asusd/`, and `/var/lib/asusd/` boundaries.

### 3.8. Runtime Performance & Non-Blocking Event Loops
* **Event-Driven Architecture:** Avoid aggressive polling loops or busy-waiting sleep cycles in background tasks (`CtrlTask`).
* **Non-Blocking Execution:** Prefer event-driven mechanisms (udev monitors, `tokio` async channels, `inotify`).
* **Allocation Inefficiencies:** Avoid unnecessary heap allocations (`String` clones, vector allocations) inside high-frequency event handling loops.

### 3.9. Internal & Workspace API Stability
* **Minimal Public API Surface:** Minimize public export signatures across internal crates (`rog-platform`, `rog-dbus`, `config-traits`).
* **Document Breaking Changes:** Explicitly document any changes to public traits, D-Bus interfaces, or library data structures.

### 3.10. Generated & Protocol Files
* **No Manual Edits to Generated Files:** Do not manually edit generated files (e.g. Slint UI bindings, D-Bus XML bindings, build manifests).
* **Source Generator Workflow:** Modifications must be made strictly through the source generator or original schema definitions.
* **Reproducibility:** Generated output must be strictly reproducible across development environments.

---

## 4. Code Integrity, Error Handling & Releases

### 4.1. No Symptom Patching
* Never swallow exceptions, return dummy fallback values, or comment out failing assertions to make tests pass.
* Trace errors back to their root cause in hardware sysfs attributes, D-Bus messages, or async task orchestration.

### 4.2. Preserving Comments & Documentation
* Preserve all existing code comments, docstrings, and architectural descriptions unless explicitly asked to modify them.
* Document public functions, trait implementations, and D-Bus methods clearly.

### 4.3. Minimal & Focused Modifications
* **Strict Task Focus:** Code and configuration modifications must always be minimal and strictly focused on the requested task.
* **Avoid Unnecessary Scope:** Do not perform extraneous refactoring, rewrite unrelated logic, or introduce changes outside the explicit boundary of the requested task.

### 4.4. Release Management, Semantic Versioning & CHANGELOG
* **Semantic Versioning (SemVer):** Follow Semantic Versioning principles (`MAJOR.MINOR.PATCH`).
* **Breaking Change Annotations:** Any breaking change to D-Bus interface paths, CLI flags, or configuration schemas must be flagged in Conventional Commit footers (`BREAKING CHANGE:`).
* **User Configuration Migration:** Ensure schema updates provide seamless backwards compatibility or migration paths for existing `/etc/asusd/` user configurations.

---

## 5. Verification & Testing Protocol

### 5.1. Verification Before Completion
* Never mark a task as complete without empirical runtime verification.
* Always execute build and check commands:
  ```bash
  cargo check --all-targets
  cargo test --all
  cargo clippy --workspace --all-targets --all-features -- -D warnings
  cargo cranky
  cargo fmt --all -- --check
  ```

### 5.2. Log Inspection
* Inspect full un-truncated logs (`journalctl`, `asusd` daemon logs, or test tracebacks) before diagnosing failures or proposing fixes.

### 5.3. Hardware-Dependent Testing & Mockability Strategy
* **Mockable Paths & Fallback Simulation:** Hardware-dependent traits and controllers must provide mockable interfaces or software fallback modes so logic can be validated on non-ASUS development machines.
* **Empirical Hardware Claim Boundary:** Never claim full physical hardware validation for model-specific features without verified logs from physical test hardware.

### 5.4. Error Path & Failure Scenario Testing
* **Failure Scenario Validation:** Include test cases and manual verification for error paths:
  - Missing sysfs/debugfs attributes or nodes.
  - Unavailable kernel drivers or ACPI WMI interfaces.
  - Corrupted or unparseable `/etc/asusd/` configuration files.
  - D-Bus daemon disconnection or client communication failures.
  - File permission or read-only filesystem errors.

### 5.5. Feature Test Coverage & Mock Strategy
* **Coverage Requirements:** New functionality must include appropriate unit or integration tests covering normal operations and edge cases.
* **Software Fallback Testing:** Hardware-specific behavior must provide mock implementations or software fallback modes when physical test hardware is unavailable.

### 5.6. Release Security & Vulnerability Audit
* **Audit Execution Before Release:** Execute `cargo audit` before releasing or merging dependency updates to ensure no known security vulnerabilities exist in workspace dependencies.

**Comment by Ghoul4500** [comment]
Working on https://github.com/OpenGamingCollective/asusctl/pull/327 we are moving away from encouraging use of AI like this. So this PR will be closed
